src/main/java/jetbrains/buildServer/auth/oauth/ServerPrincipalFactory.java

```diff
@@ -32,53 +34,94 @@ public ServerPrincipalFactory(@NotNull UserModel userModel, @NotNull UserGroupMa
-                            }
-                        }
-                    }
-                }
-            }
-            for(String group : finalGroupSet) {
-                SUserGroup userGroup = userGroupManager.findUserGroupByName(group);
-                if(userGroup != null) {
-                    userGroup.addUser(created);
-                }
+            if (syncGroups) {
+                addUserToGroups(created, applyGroupWhitelist(groupsInToken));
             }
             created.setUserProperty(PluginConstants.ID_USER_PROPERTY_KEY, user.getId());
             created.updateUserAccount(user.getId(), user.getName(), user.getEmail());
             return Optional.of(new ServerPrincipal(PluginConstants.OAUTH_AUTH_SCHEME_NAME, user.getId()));
         } else {
             LOG.info("User: " + user + " could not be found and allowCreatingNewUsersByLogin is disabled");
-            return existingPrincipal;
+            return Optional.empty();
+        }
+    }
+
+    private void removeUserFromGroups(SUser existingUser, List<String> groupsToDelete) {
+        for (String group : groupsToDelete) {
+            SUserGroup userGroup = userGroupManager.findUserGroupByName(group);
+            if (userGroup != null) {
+                userGroup.removeUser(existingUser);
+            }
+        }
+    }
+
+    private List<String> applyGroupWhitelist(List<String> groups) {
+        List<String> finalGroupList = new ArrayList<>();
+        List<String> whitelistedGroups = properties.getWhitelistedGroups();
+        if (whitelistedGroups != null) {
+            for (String group : groups) {
+                for (String whitelistedGroup : whitelistedGroups) {
+                    if (group.startsWith(whitelistedGroup)) {
+                        finalGroupList.add(group);
+                        break;
+                    }
+                }
+            }
+        }
+        return finalGroupList;
+    }
+
+    private void addUserToGroups(SUser created, List<String> finalGroupList) {
+        for (String group : finalGroupList) {
+            SUserGroup userGroup = userGroupManager.findUserGroupByName(group);
+            if (userGroup != null) {
+                userGroup.addUser(created);
+            }
         }
     }
 
     @NotNull
-    private Optional<ServerPrincipal> findExistingPrincipal(@NotNull final String userName) {
+    private Optional<SUser> findExistingUser(@NotNull final String userName) {
         try {
             final SUser user = userModel.findUserByUsername(userName, PluginConstants.ID_USER_PROPERTY_KEY);
-            return Optional.ofNullable(user).map(u -> new ServerPrincipal(PluginConstants.OAUTH_AUTH_SCHEME_NAME, user.getUsername()));
+            return Optional.ofNullable(user);
         } catch (InvalidUsernameException e) {
             // ignore it
             return Optional.empty();
         }
     }
+
+    private List<String> obtainGroupsDelta(List<String> minuend, List<String> subtrahend) {
```

Comment:
This method identifies the group names in the first arg, that are not in the second. Can't think of a better name for it.

Reply:
The method name is fine, but I had to lookup what minuend and subtrahend mean!